'use client' - flesh out usages and reference (#6362)
reactjs/react.dev: src/content/reference/react/use-client.md

<Intro>

> `'use client'` marks which components render on the client.

Reviewer: (I realize this is a draft so not reviewing in detail but wanted to note one high-level thing…)

thought: is there a way we can be clear up top that this marks the _boundary_ between server and client components? ideally people should realize that you need serializable props and that you don't need to mark every file deep down. it's great you have FancyText but I think it could be a bit clearer throughout

(obviously my intro text you replaced did not even try to clarify this! oops. what you wrote here is a clear improvement)

Author: Updated to try and highlight the concept of boundary more

### `'use client'` {/*use-client*/}

> Add `'use client'` at the top of a file to mark the module and its transitive dependencies as client code.

Reviewer 2: This use of "transitive dependencies" is correct but my brain erroneously interpreted it as "transitive dependents" instead. I wonder if we could say something like "and any other modules it imports".

Author: I think that weakens the definition though because 'any other module it imports' may be only interpreted as one level

```js {1}
'use client';

import { useState } from 'react';
import { formatDate } from './formatters';
import Link from './link';

export default function RichTextEditor(props) {
  formatDate();
  // ...
  return
    <Link />
    // ...
}
```

When a file marked with `'use client'` is imported from a Server Component, [compatible bundlers](/learn/start-a-new-react-project#bleeding-edge-react-frameworks) will treat the module import as a boundary between server-run and client-run code.

> As a dependency of `RichTextEditor`,  `formatDate` and `Link` will also be evaluated on the client regardless if their modules contain a `'use client'` directive.

Reviewer 2: regardless if -> regardless of whether

Possible follow-on sentence: Note that the same module may be evaluated on the server when imported from server code and on the client when imported from client code.

<DeepDive>

> #### Client and Server Rendering with React Server Components {/*client-and-server-rendering-with-react-server-components*/}

Reviewer 3: ```suggestion
#### Client and server rendering with React Server Components {/*client-and-server-rendering-with-react-server-components*/}
```

Titles of Deep Dive sections are not capitalized.

React's full-stack architecture vision is realized through the [React Server Components (RSC) specification](https://github.com/reactjs/rfcs/blob/main/text/0188-server-components.md).

> React Server Components lets you coordinate which components of your [render tree](/learn/understanding-your-ui-as-a-tree#the-render-tree) are rendered on the server or on the client. Server rendering is traditionally more performant and allows access to the back-end, while client rendering is necessary for adding interactivity into your components.

Reviewer 2: "traditionally" may not be the right word choice here (as opposed to actually, in the present day?) Perhaps you meant "often" or "generally".

For apps that use React Server Components, the app is server-rendered by default. `'use client'` introduces a server-client boundary in the [module dependency tree](/learn/understanding-your-ui-as-a-tree#the-module-dependency-tree), effectively creating a client-module subtree.

To better illustrate this, consider the following React Server Component app and its module dependency and render tree.

<Sandpack>

```js App.js
import FancyText from './FancyText';
import InspirationGenerator from './InspirationGenerator';
import Copyright from './Copyright';

export default function App() {
  return (
    <>
      <FancyText title text="Get Inspired App" />
      <InspirationGenerator>
        <Copyright year={2004} />
      </InspirationGenerator>
    </>
  );
}

```

```js FancyText.js
export default function FancyText({title, text}) {
  return title
    ? <h1 className='fancy title'>{text}</h1>
    : <h3 className='fancy cursive'>{text}</h3>
}
```

```js InspirationGenerator.js
'use client';

import * as React from 'react';
import inspirations from './inspirations';
import FancyText from './FancyText';

export default function InspirationGenerator({children}) {

>   const [index, setIndex] = React.useState(0);

Reviewer 2: We don't usually use 'React.---' for hooks but import them into the module namespace.

</Diagram>

We introduce the following definitions:

* **Client Components** are components in a render tree that are rendered on the client.
* **Server Components** are components in a render tree that are rendered on the server.

Working through the example app, `App`, `FancyText` and `Copyright` are all server-rendered and considered Server Components. As `InspirationGenerator.js` and its transitive dependencies are marked as client code, the component `InspirationGenerator` and its child component `FancyText` are Client Components.

<DeepDive>
#### How is `FancyText` both a Server and a Client Component? {/*how-is-fancytext-both-a-server-and-a-client-component*/}

By the above definitions, the component `FancyText` is both a Server and Client Component, how can that be?

First, let's clarify that the term "component" is not very precise. Here are just two ways "component" can be understood:

1. A "component" can refer to a **component definition**. In most cases this will be a function.

```js
// This is a definition of a component
function MyComponent() {
	return <p>My Component</p>
}
```

> 2. A "component" can also refer to a **component callsite** of its definition.

Reviewer 2: A "component" can also refer to a **component instance** created by using the component as an element within some other component.

Reviewer 2: I don't think we use the word "callsite" elsewhere. Technically speaking they aren't callsites as the component function is not called. They are technically elements but might more clearly be referred to as instances.

Author: I've heard conflicting thoughts on this from @eps1lon as I originally had them as instances. There was mention it may be confused with the runtime instance.
Before callsite, I used "use", like "component usages"

Author: I'll go back to usage for now

```js
import MyComponent from './MyComponent';

function App() {

> 	// This is a usage of a component

Author: ```suggestion
	// This is a callsite of a component
```

	return <MyComponent />;
}
```

Often, the imprecision is not important when explaining concepts, but in this case it is.

When we talk about Server or Client Components, we are referring to component callsites.

* If the component is defined in a module with a `'use client'` directive, or the component is imported and called in a Client Component, then the component callsite is a Client Component.

> * If the component is imported and called in a Server Component, then the component callsite is a Server Component.

Reviewer 2: * If the component that is _not_ defined with `use client` is imported and called in a Server Component, then that component instance is a Server Component.

Author: Maybe it should just be: "Otherwise, it is a Server Component"

Author: I've updated to that, as I use similar definition in the caveat

<Diagram name="use_client_render_tree" height={150} width={450} alt="A tree graph where each node represents a component and its children as child components. The top-level node is labelled 'App' and it has two child components 'InspirationGenerator' and 'FancyText'. 'InspirationGenerator' has two child components, 'FancyText' and 'Copyright'. Both 'InspirationGenerator' and its child component 'FancyText' are marked to be client-rendered.">A render tree illustrates component callsites.</Diagram>

Back to the question of `FancyText`, we see that the component definition does _not_ have a `'use client'` directive and it has two callsites.

> The callsite of `FancyText` as a child of `App`, marks that usage as a Server Component as root components are always Server Components. When `FancyText` is imported and called under `InspirationGenerator`, that callsite of `FancyText` is a Client Component as `InspirationGenerator` is a Client Component since it contains a `'use client'` directive.

Author: ```suggestion
The callsite of `FancyText` as a child of `App`, marks that usage as a Server Component as root components are always Server Components. When `FancyText` is imported and called under `InspirationGenerator`, that callsite of `FancyText` is a Client Component as `InspirationGenerator` contains a `'use client'` directive.
```

As a child of `InspirationGenerator`, a Client Component, why is `Copyright` a Server Component?

To clarify, `'use client'` defines the boundary between server and client code on the _module dependency tree_, not the render tree.

<Diagram name="use_client_module_dependency" height={200} width={500} alt="A tree graph with the top node representing the module 'App.js'. 'App.js' has three children: 'Copyright.js', 'FancyText.js', and 'InspirationGenerator.js'. 'InspirationGenerator.js' has two children: 'FancyText.js' and 'inspirations.js'. The nodes under and including 'InspirationGenerator.js' have a yellow background color to signify that this sub-graph is client-rendered due to the 'use client' directive in 'InspirationGenerator.js'.">
`'use client'` defines the boundary between server and client code on the module dependency tree.
</Diagram>

In the module dependency tree, we see that `App.js` imports and calls `Copyright` from the `Copyright.js` module. As `Copyright.js` does not contain a `'use client'` directive, the component callsite is rendered on the server. `App` is rendered on the server as it is the root component.

> Client Components can render Server Components because you can pass JSX as props. In this case, `InspirationGenerator` takes [JSX as children props](/learn/passing-props-to-a-component#passing-jsx-as-children) and receives `Copyright` as a `children` prop. However, the `InspirationGenerator` module never directly imports the `Copyright` module nor calls the component, all of that is done by `App`.

Author: ```suggestion
Client Components can render Server Components because you can pass JSX as props. In this case, `InspirationGenerator` receives `Copyright` as [children](/learn/passing-props-to-a-component#passing-jsx-as-children). However, the `InspirationGenerator` module never directly imports the `Copyright` module nor calls the component, all of that is done by `App`.
```

	* [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)
	* [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)
	* [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set)
	* [TypedArray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray)
* Plain [objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object), those created with [object initializers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer), with serializable properties
* Client or Server Components
* [Promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)

Notably, these are not supported:
* [Functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function) that are not exported from client-marked modules or marked with `'use server'`
* [Classes](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Classes_in_JavaScript)

> * Objects that are not an instance of [Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object), or [null-prototype objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object#null-prototype_objects)

Author: ```suggestion
* Objects that are not an instance of [Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object), or are [null-prototype objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object#null-prototype_objects)
```

* Symbols not registered globally, ex. `Symbol('my new symbol')`

#### Caveats {/*caveats*/}

* `'use client'` must be at the very beginning of a file, above any imports or other code (comments are OK). They must be written with single or double quotes, but not backticks.
* When a `'use client'` module is imported from another client-rendered module, the directive has no effect.
* When a component module contains a `'use client'` directive, any usage of that component is guaranteed to be a Client Component. However, a component can still be evaluated on the client even if it does not have a `'use client'` directive.

> 	* A component usage is considered a Client Component if it is defined in module with `'use client'` directive or it is a transitive dependency of a module that contains a `'use client'` directive. Otherwise, it is a Server Component.

Reviewer 2: or it is -> or when it is

and clarify that just importing it from one client component doesn't taint it into running on the client when some other server component imports it.

## Usage {/*usage*/}

### Building with interactivity and state {/*building-with-interactivity-and-state*/}

<Sandpack>

```js App.js
'use client';

import {useState} from 'react';

export default function Counter({initialValue = 0}) {
	const [countValue, setCountValue] = useState(initialValue);
	const increment = () => setCountValue(countValue + 1);
	const decrement = () => setCountValue(countValue - 1);
	return (<>
		<h2>Count Value: {countValue}</h2>
		<button onClick={increment}>+1</button>
		<button onClick={decrement}>-1</button>
	</>);
}
```

</Sandpack>

As `Counter` requires both the `useState` hook and event handlers to increment or decrement the value, this component must be a Client Component and will require a `'use client'` directive at the top.

In contrast, a component that renders UI without interaction will not need to be a Client Component.

```js
import {getCounterValueFromFile} from 'fs-utils';
import Counter from './Counter';

export default async function CounterContainer() {
  const initialValue = await getCounterValueFromFile();
  return <Counter initialValue={initialValue} />
}
```

> For example, the parent component of the Client Component `CounterContainer` does not require `'use client'` as it is not interactive and does not use state. To be more specific, `CounterContainer` must be a Server Component as it reads from the local file system on the server. Notably, Server Components support asynchronous rendering, which is not supported for Client Components.

Reviewer 2: To be more specific, -->  In addition, / Moreover,  —— this is just a further fact rather than an increase of specificity.

Notably, Server Components support asynchronous rendering, which is not supported for Client Components. ---> (This is possible because Server Components, unlike Client Components, can be async functions.)

I would say "can be async functions" because this is more specific and actionable than "support asynchronous rendering" which is an abstract concept that could refer to concurrent rendering or who knows whatw.

There are also components that don't use any server or client-only features and can be agnostic to where they render. `FancyText` is an example of such a component.

```js
export default function FancyText({title, text}) {
  return title
    ? <h1 className='fancy title'>{text}</h1>
    : <h3 className='fancy cursive'>{text}</h3>
}
```

> In this case, it is discouraged to use the `'use client'` directive as it pre-maturely forces all component usages of `FancyText` to be rendered on the client, which comes at a performance cost. As demonstrated in the earlier Inspirations app example, `FancyText` is used as both a Server or Client Component, depending on where it is imported and used.

Reviewer 2: ```suggestion
In this case, it is discouraged to use the `'use client'` directive as it prematurely forces all component usages of `FancyText` to be rendered on the client, which comes at a performance cost. As demonstrated in the earlier Inspirations app example, `FancyText` is used as both a Server or Client Component, depending on where it is imported and used.
```

```js
'use client';

import {useRef, useEffect} from 'react';

export default function Circle() {
  const ref = useRef(null);
  useLayoutEffect(() => {
    const canvas = ref.current;
    const context = canvas.getContext('2d');
    context.beginPath();
    context.arc(100, 75, 50, 0, 2 * Math.PI);
    context.stroke();
  });
  return <canvas ref={ref} />
}
```

### Using React libraries {/*using-react-libraries*/}

Often in a React app, you'll leverage third-party libraries to handle common UI patterns or logic.

These libraries may rely on component Hooks or client APIs. In these cases, you'll need to ensure you're using these libraries in Client Components. Depending on the nature of these libraries, this may mean adding a `'use client'` near the top of your module dependency tree – marking the majority of your React app as client-rendered.

Libraries that use any of the following React APIs must be marked as client-run:
* [createContext](/reference/react/createContext)

> * [`react`](/reference/react) and [`react-dom`](/reference/react-dom/hooks) Hooks

Author: I forget, I thought `use` wasn't recommended on the server? cc @mattcarrollcode

> * [`react`](/reference/react) and [`react-dom`](/reference/react-dom/hooks) Hooks

Reviewer: use and useId can be used sparingly in server components. But async/await is better if all you want to do is unwrap a promise. It's really only if you need useId then you can't do async/await, and then you need use() if you want to "await" in the same component.

> * [`react`](/reference/react) and [`react-dom`](/reference/react-dom/hooks) Hooks

Author: Updated to include those exceptions

  ];
```

```css
.fancy {
  font-family: 'Georgia';
}
.title {
  color: #007AA3;
  text-decoration: underline;
}
.cursive {
  font-style: italic;
}
.small {
  font-size: 10px;
}
```

</Sandpack>

<Diagram name="use_client_module_dependency" height={250} width={545} alt="A tree graph with the top node representing the module 'App.js'. 'App.js' has three children: 'Copyright.js', 'FancyText.js', and 'InspirationGenerator.js'. 'InspirationGenerator.js' has two children: 'FancyText.js' and 'inspirations.js'. The nodes under and including 'InspirationGenerator.js' have a yellow background color to signify that this sub-graph is client-rendered due to the 'use client' directive in 'InspirationGenerator.js'.">
`'use client'` segments the module dependency tree of the React Server Components app to mark `InspirationGenerator.js` and any of its dependencies as client-rendered.
</Diagram>

In the module dependency tree of the example app, the `'use client'` directive in `InspirationGenerator.js` marks that module and all of its transitive dependencies as client modules. It creates a subtree of client modules with `InspirationGenerator.js` as the root.

> Irregardless if a module exports a component, like `inspirations.js`, if a module is part of that client subtree, then it will be evaluated on the client.

Author: I think I'll just remove it as I've clarified earlier that code and component are both marked as client evaluated.
